Schnorr & Taproot extension of Bitcoin Core TestFramework library

jachiang opened the pull request:
This is the Schnorr & Taproot library extension of the Bitcoin Core Python TestFramework I have been working on during my residency. Although the TestFramework is intended to provide functional test coverage for the Bitcoin Core project, it can also serve as a convenient Schnorr/Taproot prototyping tool for developers.

The goal of this project was to provide a toolkit for users to quickly prototype new Bitcoin contracting proposals with Schnorr & Taproot, so that the community can quickly get familiar and provide constructive feedback.

**_Schnorr extensions_**
 - [x] Musig signature aggregation
 - [ ] Adaptor Signature generation (Removed from this PR, out of scope for Optech)
 - [ ] Discrete Log Contract key generation (Removed, out of scope for Optech)

**_Taproot output classes (Segwit V1)_**
 - [x] Taptree & constructor methods (huffman construtor with weighted tapleafs)
 - [x] Tapscript descriptors (see below)

**_Minscript classes_** (can be composed to (correctly) express tapscript descriptors)
For example:
 - [x] `pkolder(key, time) = and_v(v:pk(key),older(time))`
 - [x] `csahash(n, keys, hash) = and_v(v:thresh_csa,ripemd160)`
 - [x] `csahasholder(n,keys,hash,time) = and_v(v:thresh_csa ,and_v(v:ripemd160(hash), older(time)))`

**_Utility Classes/Methods_**
  - [x] TestWrapper class wraps functionality of BitcoinTestFramework for external usage.
  - [x] Refactor BitcoinTestFramework to encapsulate code in `main` into `setup` and `shutdown`, in order to prevent current code duplication in TestWrapper. I will also create a PR in master for this once completed.

_**Future work**_
  - [ ] Complete miniscript language support.
  - [ ] Policy to Miniscript compiler for Taproot outputs (Policy language can provide flexible composition of spending conditions for taproot outputs. The compiler will separate policy sub-expressions in different tapleafs for user-defined privacy.)

Your feedback and review is greatly appreciated.

I am currently documenting the library in form of interactive jupyter notebook files in [bitcoinops/taproot](https://github.com/bitcoinops/taproot).

jachiang committed:
[test] Add Taptree, Tapleaf classes and generator methods.

Added miniscript class and expressions.

Added subset of miniscript, which can be used to construct/satisfy
scripts in a composable manner.

Add miniscript to TapLeaf descriptors.

TapLeaf class now deserializes/serializes descriptors with miniscript
nodes. This facilitates construction of new tapleaf descriptors due to
composability. New TapLeaf descriptors added containing time/hash
expressions.

Added TapLeaf test for pk descriptor types.

Test constructs and spends scriptpaths of taptree which is constructed
from various pk tapscript descriptor types containing pk, hash and time
expressions.

Added TapLeaf test for csa descriptor types.

Test constructs and spends scriptpaths of taptree which is constructed
from various csa (checksigadd) tapscript descriptor types containing pk,
hash and time expressions.

jachiang committed:
[test] Add add/multiply methods to ECPubkey and ECKey classes.

These methods are required for the multiplication/tweaking of
private/public key pairs. Keypairs are multiplied by a challenge factor
in musig signatures. Keypairs are tweaked in both discrete log contracts
and adaptor signatures.

jachiang committed:
[test] Add schnorr method which allows signing with a valid, predetermined Schnorr nonce.

jachiang committed:
[test] Add MuSig signature support

MuSig scheme is based on the interactive MuSig proposal, which is not
formalized into a BIP proposal at the time of this writing.

jachiang committed:
[test] Refactor BitcoinTestFramework main function.

In order to wrap the BitcoinTestFramework class in a user-defined class,
which can be started and shutdown, the code in BitcoinTestFramework main
has been encapsulated in respective functions. Additionally, test-run
specific state in temp-directory and NetworkThread is now reset after
shutdown.

jnewbery commented:
Hi James. Thanks for removing the adaptor signature/DLC stuff. There are still a couple of things more to do before this is ready for review:

- squash the 'fixup' commits. (eg remove the _Removed Whitespaces. Edited var name & comments._ and _Min nVersion corrected to 1 in tests._ commits)
- make the framework changes in the same (or consecutive) commits to the tests in which they're first used.

Could you also open a PR just for the _Add TestFramework reset required by TestWrapper._ and _Reorganization of BitcoinTestFramework main method._ commits please (since we'd want to get those merged into bitcoin/bitcoin).

jachiang commented:
> * squash the 'fixup' commits. (eg remove the _Removed Whitespaces. Edited var name & comments._ and _Min nVersion corrected to 1 in tests._ commits)
> * make the framework changes in the same (or consecutive) commits to the tests in which they're first used.
> 
> Could you also open a PR just for the _Add TestFramework reset required by TestWrapper._ and _Reorganization of BitcoinTestFramework main method._ commits please (since we'd want to get those merged into bitcoin/bitcoin).

Hi @jnewbery - Many thanks for the preliminary review. I have addressed your comments as suggested.

jnewbery commented on test/functional/test_framework/key.py:
```diff
@@ -356,6 +356,26 @@ def tweak_add(self, tweak):
         return ret
 
+    def mul(self, data):
+        assert(self.valid)
```
nit: I know this already occurs throughout this module, but in Python, [assert is a statement, not a function](https://stackoverflow.com/questions/13390401/design-of-python-why-is-assert-a-statement-and-not-a-function), so this should be `assert self.valid`. No need to change it here, but this would be better without the parens.

jnewbery commented:
I've force pushed a change to fix some general best practices:

- wrapped commit logs at 120 chars
- removed unused python imports and sorted imports according to [PEP 8](https://www.python.org/dev/peps/pep-0008/).

original branch is here: https://github.com/jnewbery/bitcoin/tree/2019-09-optech-taproot-master-original

jnewbery commented:
force-pushed some more updates:

- changed file permissions so test files are runnable (chmod 755)
- renamed test files to be `feature_` instead of `framework_`
- added boilerplate licence text and doc strings
- fixed all [flake8](http://flake8.pycqa.org/en/latest/) warnings

jnewbery commented on test/functional/test_framework/key.py:
```diff
@@ -356,6 +356,26 @@ def tweak_add(self, tweak):
         ret.compressed = self.compressed
         return ret
 
+    def mul(self, data):
```
Add a docstring commenting this function (ie that this multiplies a public key by a scalar).

jachiang replied:
Done.

jnewbery commented on test/functional/test_framework/key.py:
```diff
@@ -356,6 +356,26 @@ def tweak_add(self, tweak):
+        ret.compressed = self.compressed
+        return ret
+
+    def add(self, pk):
```
Add a docstring commenting this function (ie that this adds two pubkeys)

jachiang replied:
Done.

jnewbery commented on test/functional/test_framework/key.py:
```diff
@@ -356,6 +356,26 @@ def tweak_add(self, tweak):
+        ret.compressed = self.compressed
+        return ret
+
+    def add(self, pk):
```
Did you consider using the [`__add__`](https://docs.python.org/3/reference/datamodel.html#object.__add__) method so users can use the more natural notation:

```
a = ECPubKey()
b = ECPubKey()
c = a + b
```

than:

```
a = ECPubKey()
b = ECPubKey()
c = a.add(b)
```

(same comment for other `add()` and `mul()` functions in this commit)

jachiang replied:
Done. I kept add/mul for key operations for scalar bytes (e.g. tweaking a key with hash digest), but these are just wrappers for +/- to minimise code duplication.

jnewbery commented on test/functional/test_framework/key.py:
```diff
@@ -380,6 +400,26 @@ def get_bytes(self):
 
+    def add(self, data):
+        """Add key to a scalar"""
+        secret = int.from_bytes(data, 'big')
```
I don't think the variable name `secret` is appropriate here. Perhaps `summand`?

jnewbery commented on test/functional/test_framework/key.py:
```diff
@@ -431,6 +471,41 @@ def sign_schnorr(self, msg):
         e = int.from_bytes(hashlib.sha256(R[0].to_bytes(32, 'big') + self.get_pubkey().get_bytes() + msg).digest(), 'big') % SECP256K1_ORDER
         return R[0].to_bytes(32, 'big') + ((k + e*self.secret) % SECP256K1_ORDER).to_bytes(32, 'big')
 
+    def sign_schnorr_with_nonce(self, msg, k_key):
```
This is almost identical to the code above in `sign_schnorr()`. Can you just add an optional parameter `nonce=None` to `sign_schnorr()` instead of copy-pasting a lot of code?

jnewbery commented on test/functional/test_framework/key.py:
```diff
@@ -431,6 +471,41 @@ def sign_schnorr(self, msg):
+        assert(self.valid)
+        assert(self.compressed)
+        assert(len(msg) == 32)
+        if k_key == None:
```
`if k_key is None` is preferred over `if k_key == None`

jnewbery commented on test/functional/test_framework/key.py:
```diff
@@ -444,3 +519,68 @@ def tweak_add(self, tweak):
         return ret
+
+def generate_schnorr_nonce():
+    """Generate a random valid Schnorr nonce."""
```
This should be "valid bip-schnorr nonce" rather than "valid Schnorr nonce", and should explain what makes it valid (that it is a quadratic residue). You could also refer to the signing algorithm in bip-schnorr (https://github.com/sipa/bips/blob/bip-schnorr/bip-schnorr.mediawiki#Signing) and say that this isn't the deterministic algorithm used there.

jnewbery commented on test/functional/test_framework/key.py:
```diff
@@ -431,6 +471,41 @@ def sign_schnorr(self, msg):
+        e = int.from_bytes(hashlib.sha256(R[0].to_bytes(32, 'big') + self.get_pubkey().get_bytes() + msg).digest(), 'big') % SECP256K1_ORDER
+        return R[0].to_bytes(32, 'big') + ((k + e*self.secret) % SECP256K1_ORDER).to_bytes(32, 'big')
+
+    def sign_musig(self, k_key, R_negated, R_musig, P_musig, msg):
```
I think that this (and `aggregate_schnorr_nonces()`, `generate_musig_key()` and `aggregate_musig_signatures()` below) should all move to their own indvidual musig.py module.

jnewbery replied:
Is it possible to remove the `R_negated` parameter here, and just have the individual signers update their `r` values to negate them when necessary?

jnewbery commented on test/functional/test_framework/key.py:
```diff
@@ -444,3 +519,68 @@ def tweak_add(self, tweak):
+def generate_musig_key(PubkeyList):
+    """Aggregate individually generated public keys."""
+    pkv = [int.from_bytes(key.get_bytes()[1:], 'big') for key in PubkeyList]
+    pkv.sort()
```
nit: join with line above.

jnewbery commented on test/functional/test_framework/key.py:
```diff
@@ -444,3 +519,68 @@ def tweak_add(self, tweak):
+
+def generate_musig_key(PubkeyList):
+    """Aggregate individually generated public keys."""
+    pkv = [int.from_bytes(key.get_bytes()[1:], 'big') for key in PubkeyList]
```
nit: no need to use such short variable names. I think `pubkeys` is a better name here, and `pubkey` instead of `px`

jnewbery commented on test/functional/test_framework/key.py:
```diff
@@ -444,3 +519,68 @@ def tweak_add(self, tweak):
+    ret.compressed = True
+    return ret, negated
+
+def generate_musig_key(PubkeyList):
```
nit: convention is to use `snake_case` for variables, functions and arguments, and `CamelCase` for class names.

jnewbery commented on test/functional/test_framework/key.py:
```diff
@@ -444,3 +519,68 @@ def tweak_add(self, tweak):
+    k_key.set(k.to_bytes(32, 'big'), True)
+    return k_key
+
+def aggregate_schnorr_nonces(Rv):
```
nit: change argument name to `nonces`

jnewbery commented on test/functional/test_framework/key.py:
```diff
@@ -444,3 +519,68 @@ def tweak_add(self, tweak):
+
+def aggregate_schnorr_nonces(Rv):
+    """Construct aggregated musig nonce from individually generated nonces."""
+    for idx, R in enumerate(Rv):
```
nit: consider using a map-reduce here:

```
from functools import reduce
R_agg = reduce(lambda x, y: SECP256K1.add(x, y), map(lambda x: x.p, Rv))
```

actually, since you've already defined a way to add `ECPubKey`s, you don't even need the map:

```
R_agg = reduce(lambda x, y: x.add(y), Rv).p
```

If you define `__add__` for `ECPubKey`, then this just becomes:

```
R_agg = sum(Rv).p
```

edit: since you're returning a `ECPubKey` object, just set `R_agg` as a `ECPubKey`:

```
R_agg = reduce(lambda x, y: x.add(y), Rv)
```

jachiang replied:
Very nice!

jnewbery commented on test/functional/test_framework/key.py:
```diff
@@ -444,3 +519,68 @@ def tweak_add(self, tweak):
+        musig_c[key] = hashlib.sha256(Lh + key.get_bytes()[1:]).digest()
+        c = int.from_bytes(musig_c[key],'big')
+        pc = SECP256K1.mul([(key.p, c)])
+        if key.get_bytes() == PubkeyList[0].get_bytes():
```
Again, I suggest using a `reduce` pattern here.

jachiang replied:
Done.

jnewbery commented on test/functional/test_framework/key.py:
```diff
@@ -444,3 +519,68 @@ def tweak_add(self, tweak):
+    return ret, negated
+
+def generate_musig_key(PubkeyList):
+    """Aggregate individually generated public keys."""
```
This comment could be improved by documenting what is being returned.

jnewbery commented on test/functional/test_framework/key.py:
```diff
@@ -444,3 +519,68 @@ def tweak_add(self, tweak):
+    for idx, sig in enumerate(sigs):
+        s = sig[32:]
+        r_x_data = sig[:32]
+        assert(len(s)==32 and len(r_x_data)==32)
```
`len(s)` is guaranteed to be 32 by construction. It would be better to assert that `len(sig) == 64` at the top.

jnewbery commented on test/functional/test_framework/key.py:
```diff
@@ -444,3 +519,68 @@ def tweak_add(self, tweak):
+        s = sig[32:]
+        r_x_data = sig[:32]
+        assert(len(s)==32 and len(r_x_data)==32)
+        s_agg = int.from_bytes(s, 'big') if idx == 0 else (s_agg + int.from_bytes(s, 'big')) % SECP256K1_ORDER
```
Use `sum` or `reduce` here (and for the `r` aggregation below)

jnewbery reviewed (commented):
I've reviewed all the changes in key.py. I haven't reviewed `script.py` or the individual test scripts.

jachiang commented:
@jnewbery - Many thanks for your detailed review. I have incorporated all of your feedback in the latest push.

jnewbery commented:
I've force-pushed a branch that squashes the latest changes into the appropriate commits, and also adds the tests to `test_runner.py`

jnewbery committed:
add/mul method nits

jnewbery committed:
schnorr method nits

jnewbery committed:
musig nits

jnewbery commented on test/functional/test_framework/musig.py:
This is duplicating code from `sign_schnorr` in key.py. Can you just call through to that method to avoid the duplication?

jnewbery commented on test/functional/test_framework/musig.py:
Can you document what the arguments are in this docstring?

jnewbery reviewed (commented):
I've pushed three commits containing nits for the add/mul commit, the schnorr commit and the musig commit.

@jachiang - can you review those commits and squash them into the preceding commits if you're happy with them?

I've also left a couple of small comments on musig.py

jnewbery committed:
Add subtract method

jnewbery commented:
@jachiang - this code is still undergoing active development in the https://github.com/bitcoinops/taproot-workshop repo. I suggest we close this PR until that's settled down, and then perhaps you can re-open against https://github.com/bitcoin/bitcoin when there's a schnorr/taproot PR?

jachiang commented:
> @jachiang - this code is still undergoing active development in the https://github.com/bitcoinops/taproot-workshop repo. I suggest we close this PR until that's settled down, and then perhaps you can re-open against https://github.com/bitcoin/bitcoin when there's a schnorr/taproot PR?

Sounds good @jnewbery! Many thanks again for your extensive review on this PR.
